### PR TITLE
Ensure that the connecting Elastic Agent is a supported version.

### LINKED ISSUE #238: Validate fleet-server compatibility in Elastic Agent
Currently Elastic Agent talks to Kibana and verifies the version. In the future this will fleet-server. For Elastic Agent there are 2 compatibility parts:

* Standalone: Compatibility with Elasticsearch
* Managed: Compatibility with fleet-server

As the compatibility with Elasticsearch in the managed mode is enforced by fleet-server (https://github.com/elastic/fleet-server/issues/167) this should not be an issue. In general the compatibility looks as following:

`Elastic Agent <= fleet-server <= Elasticsearch`

The bugfix release might differ. The last minor or Elastic Agent in a major should be compatible with the first minor of the next major of fleet-server. Same for Elasticsearch. This means fleet-server / Elasticsearch must be upgraded first.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds checks to the enrollment and checkin API to ensure that the `User-Agent` header is correct for an Elastic Agent. This verifies that the version of the Elastic Agent is a support version.

This does allow an Elastic Agent to be on a higher patched version than the Fleet Server, but not higher on the MAJOR.MINOR.

## Why is it important?

Older Elastic Agents will not work correctly with new Fleet Server, so they need to not connect to the Fleet Server.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

## Related issues

- Closes #238

### CODE REVIEW

File: cmd/fleet/handleCheckin.go [hunk 1]
```diff
@@ -79,6 +81,7 @@ type CheckinT struct {
 }
 
 func NewCheckinT(
+	verCon version.Constraints,
```
Comment (aleksmaus) [inline cmd/fleet/handleCheckin.go:84]: are we always going to have the version constraint? 
if not not then make it option function?
nit: the number of params in the constructor "is too darn high" :-)

File: cmd/fleet/userAgent.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// maximizePatch turns the version into a string that has the patch value set to the maximum integer.
+//
+// Used to allow the Elastic Agent to be at a higher patch version than the Fleet Server, but require that the
+// Elastic Agent is not higher in MAJOR or MINOR.
+func maximizePatch(ver *version.Version) string {
+	segments := ver.Segments()
+	if len(segments) > 2 {
+		segments = segments[:2]
+	}
+	segments = append(segments, int(^uint(0)>>1)) // max int
```
Comment (aleksmaus) [inline cmd/fleet/userAgent.go:47]:
```
math.MaxInt32
```
?

File: cmd/fleet/userAgent.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+// validateUserAgent validates that the User-Agent of the connecting Elastic Agent is valid and that the version is
+// supported for this Fleet Server.
+func validateUserAgent(r *http.Request, verConst version.Constraints) error {
+	userAgent := r.Header.Get("User-Agent")
+	if userAgent == "" {
+		return ErrInvalidUserAgent
+	}
+	userAgent = strings.ToLower(userAgent)
+	if !strings.HasPrefix(userAgent, "elastic agent ") {
```
Comment (aleksmaus) [inline cmd/fleet/userAgent.go:63]:
```
const prefix = "elastic agent "
```
?
could be local to func
cause it is used in two places

**Review by aleksmaus — APPROVED** [review]
couple of nits, otherwise lgtm

File: cmd/fleet/handleCheckin.go (hunk 1 above)
Comment (blakerouse) [reply cmd/fleet/handleCheckin.go:84]: yes we should always have it

File: cmd/fleet/userAgent.go (hunk 2 above)
Comment (blakerouse) [reply cmd/fleet/userAgent.go:47]: didn't know about it, will change to that

File: cmd/fleet/userAgent.go (hunk 3 above)
Comment (blakerouse) [reply cmd/fleet/userAgent.go:63]: will do

File: cmd/fleet/handleCheckin.go [hunk 4]
```diff
@@ -124,6 +128,11 @@ func (ct *CheckinT) _handleCheckin(w http.ResponseWriter, r *http.Request, id st
 		return err
 	}
 
+	err = validateUserAgent(r, ct.verCon)
+	if err != nil {
+		return err
+	}
+
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply cmd/fleet/handleCheckin.go:135]: Yeah I put it before the metric in both check-in and enroll. Can change in follow up if needed.

**Pull request merged**